org-code-javabuilder/lib/src/main/java/org/code/javabuilder/JavaRunner.java

```diff
@@ -61,4 +54,43 @@ public void runCode()
       throw new InternalFacingException("Error closing urlClassLoader: " + e, e);
     }
   }
+
+  /**
+   * Finds the main method in the set of files in fileManager if it exists.
+   *
+   * @param urlClassLoader class loader pointing to location of compiled classes
+   * @return the main method if it is found
+   * @throws UserFacingException if there is an issue loading a class
+   * @throws UserInitiatedException if there is more than one main method or no main method
+   */
+  public Method findMainMethod(URLClassLoader urlClassLoader)
+      throws UserFacingException, UserInitiatedException {
+
+    Method mainMethod = null;
+    List<ProjectFile> fileList = this.fileManager.getFiles();
+    for (int i = 0; i < fileList.size(); i++) {
+      ProjectFile file = fileList.get(i);
+      try {
+        Method tempMain =
+            urlClassLoader.loadClass(file.getClassName()).getDeclaredMethod("main", String[].class);
+        if (mainMethod != null) {
+          throw new UserInitiatedException(
+              "Your code can only have one main method. We found at least two classes with main methods.");
+        } else {
+          mainMethod = tempMain;
+        }
+      } catch (ClassNotFoundException e) {
+        // this should be caught earlier in compilation
+        throw new UserFacingException(
+            "We hit an error on our side while running your program. Try Again", e);
+      } catch (NoSuchMethodException e) {
```

Comment:
Using an exception for normal code flow is a bit confusing semantically - i.e. I'd normally expect to see some sort of error here. It also can have performance impacts, depending on the JVM and compiler. It's also Looks like `Class` also has `getDeclaredMethods` which returns an array of methods. That could prevent this exception from being hit after each non-main class is searched.

Reply:
good find!